File: cmake/MverPhaseMetrics.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Runtime.InteropServices;

public sealed class MverPhaseCandidate
{
    public int NativeOffsetFrames { get; set; }
    public double NativeOffsetSeconds { get; set; }
    public int OverlapFrames { get; set; }
    public double FullSimilarity { get; set; }
    public double FaceSimilarity { get; set; }
    public double HairSimilarity { get; set; }
    public double HandSimilarity { get; set; }
    public double FullMotionSimilarity { get; set; }
    public double FaceMotionSimilarity { get; set; }
    public double HairMotionSimilarity { get; set; }
    public double HandMotionSimilarity { get; set; }
}

public static class MverPhaseMetrics
{
    private sealed class Frame
    {
        public int Width;
        public int Height;
        public int GridWidth;
        public int GridHeight;
        public byte[] Bgr;
    }

    private sealed class Score
    {
        public readonly long[] Absolute = new long[4];
        public readonly long[] Samples = new long[4];
    }

    private sealed class MotionScore
    {
        public readonly double[] Dot = new double[4];
        public readonly double[] FirstSquared = new double[4];
        public readonly double[] SecondSquared = new double[4];
    }

    private const int PixelStep = 4;

    private static Frame Load(string path)
    {
        using (var source = new Bitmap(path))
        using (var bitmap = new Bitmap(source.Width, source.Height,
            PixelFormat.Format32bppArgb))
        {
            using (Graphics graphics = Graphics.FromImage(bitmap))
            {
                graphics.CompositingMode = CompositingMode.SourceCopy;
                graphics.DrawImageUnscaled(source, 0, 0);
            }
            var rect = new Rectangle(0, 0, bitmap.Width, bitmap.Height);
            BitmapData locked = bitmap.LockBits(rect, ImageLockMode.ReadOnly,
                PixelFormat.Format32bppArgb);
            int stride = Math.Abs(locked.Stride);
            var sourceBytes = new byte[stride * bitmap.Height];
            Marshal.Copy(locked.Scan0, sourceBytes, 0, sourceBytes.Length);
            bitmap.UnlockBits(locked);
            int gridWidth = (bitmap.Width + PixelStep - 1) / PixelStep;
            int gridHeight = (bitmap.Height + PixelStep - 1) / PixelStep;
            var bgr = new byte[gridWidth * gridHeight * 3];
            for (int gy = 0; gy < gridHeight; ++gy)
            for (int gx = 0; gx < gridWidth; ++gx)
            {
                int sourceOffset = Math.Min(bitmap.Height - 1, gy * PixelStep) * stride +
                    Math.Min(bitmap.Width - 1, gx * PixelStep) * 4;
                int outputOffset = (gy * gridWidth + gx) * 3;
                bgr[outputOffset] = sourceBytes[sourceOffset];
                bgr[outputOffset + 1] = sourceBytes[sourceOffset + 1];
                bgr[outputOffset + 2] = sourceBytes[sourceOffset + 2];
            }
            return new Frame { Width = bitmap.Width, Height = bitmap.Height,
                GridWidth = gridWidth, GridHeight = gridHeight, Bgr = bgr };
        }
    }

    private static Frame[] LoadAll(string[] paths)
    {
        var result = new Frame[paths.Length];
        for (int i = 0; i < paths.Length; ++i) result[i] = Load(paths[i]);
        for (int i = 1; i < result.Length; ++i)
            if (result[i].Width != result[0].Width ||
                result[i].Height != result[0].Height)
                throw new InvalidOperationException("Frame dimensions differ");
        return result;
    }

    private static bool InRegion(int region, int x, int y, int width, int height)
    {
        double nx = x / (double)width;
        double ny = y / (double)height;
        if (region == 0) return true;
        if (region == 1) return nx >= .27 && nx <= .74 && ny >= .24 && ny <= .67;
        if (region == 2) return ny <= .36 ||
            ((nx <= .30 || nx >= .71) && ny <= .70);
        return nx >= .16 && nx <= .58 && ny >= .58 && ny <= .92;
    }

    private static void AddPair(Score score, Frame first, Frame second,
        int spatialStride)
    {
        if (first.Width != second.Width || first.Height != second.Height)
            throw new InvalidOperationException("Frame dimensions differ");
        for (int gy = 0; gy < first.GridHeight; gy += spatialStride)
        for (int gx = 0; gx < first.GridWidth; gx += spatialStride)
        {
            int offset = (gy * first.GridWidth + gx) * 3;
            int firstMaximum = Math.Max(first.Bgr[offset],
                Math.Max(first.Bgr[offset + 1], first.Bgr[offset + 2]));
            int secondMaximum = Math.Max(second.Bgr[offset],
                Math.Max(second.Bgr[offset + 1], second.Bgr[offset + 2]));
            if (firstMaximum <= 4 && secondMaximum <= 4) continue;
            int absolute = Math.Abs(first.Bgr[offset] - second.Bgr[offset]) +
                Math.Abs(first.Bgr[offset + 1] - second.Bgr[offset + 1]) +
                Math.Abs(first.Bgr[offset + 2] - second.Bgr[offset + 2]);
            int x = gx * PixelStep;
            int y = gy * PixelStep;
            for (int region = 0; region < 4; ++region)
            {
                if (!InRegion(region, x, y, first.Width, first.Height)) continue;
                score.Absolute[region] += absolute;
                score.Samples[region]++;
            }
        }
    }

    private static double Similarity(Score score, int region)
    {
        return score.Samples[region] == 0 ? 0.0 : 100.0 *
            (1.0 - score.Absolute[region] / (score.Samples[region] * 3.0 * 255.0));
    }

    private static void AddMotionPair(MotionScore score,
        Frame firstPrevious, Frame first, Frame secondPrevious, Frame second,
        int spatialStride)
    {
        for (int gy = 0; gy < first.GridHeight; gy += spatialStride)
        for (int gx = 0; gx < first.GridWidth; gx += spatialStride)
        {
            int offset = (gy * first.GridWidth + gx) * 3;
            int x = gx * PixelStep;
            int y = gy * PixelStep;
            for (int channel = 0; channel < 3; ++channel)
            {
                double firstDelta = first.Bgr[offset + channel] -
                    firstPrevious.Bgr[offset + channel];
                double secondDelta = second.Bgr[offset + channel] -
                    secondPrevious.Bgr[offset + channel];
                if (firstDelta == 0.0 && secondDelta == 0.0) continue;
                for (int region = 0; region < 4; ++region)
                {
                    if (!InRegion(region, x, y, first.Width, first.Height)) continue;
                    score.Dot[region] += firstDelta * secondDelta;
                    score.FirstSquared[region] += firstDelta * firstDelta;
                    score.SecondSquared[region] += secondDelta * secondDelta;
                }
            }
        }
    }

    private static double MotionSimilarity(MotionScore score, int region)
    {
        double denominator = Math.Sqrt(score.FirstSquared[region] *
            score.SecondSquared[region]);
        if (denominator <= 0.0) return 0.0;
        double correlation = Math.Max(-1.0, Math.Min(1.0,
            score.Dot[region] / denominator));
        return 50.0 * (correlation + 1.0);
    }

    private static MverPhaseCandidate Candidate(Score score, MotionScore motion,
        int offset, int overlap, int intervalMilliseconds)
    {
        return new MverPhaseCandidate {
            NativeOffsetFrames = offset,
            NativeOffsetSeconds = offset * intervalMilliseconds / 1000.0,
            OverlapFrames = overlap,
            FullSimilarity = Similarity(score, 0),
            FaceSimilarity = Similarity(score, 1),
            HairSimilarity = Similarity(score, 2),
            HandSimilarity = Similarity(score, 3),
            FullMotionSimilarity = MotionSimilarity(motion, 0),
            FaceMotionSimilarity = MotionSimilarity(motion, 1),
            HairMotionSimilarity = MotionSimilarity(motion, 2),
            HandMotionSimilarity = MotionSimilarity(motion, 3)
        };
    }

    private static MverPhaseCandidate Measure(Frame[] mver, Frame[] native,
        int offset, int intervalMilliseconds, int spatialStride)
    {
        int mverStart = Math.Max(0, -offset);
        int nativeStart = Math.Max(0, offset);
        int overlap = Math.Min(mver.Length - mverStart,
            native.Length - nativeStart);
        var score = new Score();
        var motion = new MotionScore();
        for (int frame = 0; frame < overlap; frame += 2)
            AddPair(score, mver[mverStart + frame],
                native[nativeStart + frame], spatialStride);
        for (int frame = 1; frame < overlap; ++frame)
            AddMotionPair(motion, mver[mverStart + frame - 1],
                mver[mverStart + frame], native[nativeStart + frame - 1],
                native[nativeStart + frame], spatialStride);
        return Candidate(score, motion, offset, overlap, intervalMilliseconds);
    }

    public static MverPhaseCandidate[] Search(string[] mverPaths,
        string[] nativePaths, int maxLagFrames, int minimumOverlapFrames,
        int intervalMilliseconds)
    {
        if (maxLagFrames < 0 || minimumOverlapFrames < 2 ||
            mverPaths.Length < minimumOverlapFrames ||
            nativePaths.Length < minimumOverlapFrames)
            throw new ArgumentException("Insufficient frames for lag analysis");
        Frame[] mver = LoadAll(mverPaths);
        Frame[] native = LoadAll(nativePaths);
        var candidates = new List<MverPhaseCandidate>(maxLagFrames * 2 + 1);
        for (int offset = -maxLagFrames; offset <= maxLagFrames; ++offset)
        {
            MverPhaseCandidate candidate = Measure(mver, native, offset,
                intervalMilliseconds, 3);
            if (candidate.OverlapFrames >= minimumOverlapFrames)
                candidates.Add(candidate);
        }
        if (candidates.Count == 0)
            throw new ArgumentException("No lag candidate has enough overlap");
        candidates.Sort(delegate(MverPhaseCandidate left, MverPhaseCandidate right) {
            return right.FullMotionSimilarity.CompareTo(left.FullMotionSimilarity);
        });
        var fineCandidates = new List<MverPhaseCandidate>();
        var offsets = new HashSet<int>();
        for (int index = 0; index < Math.Min(7, candidates.Count); ++index)
            offsets.Add(candidates[index].NativeOffsetFrames);
        offsets.Add(0);
        foreach (int offset in offsets)
        {
            MverPhaseCandidate candidate = Measure(mver, native, offset,
                intervalMilliseconds, 1);
            if (candidate.OverlapFrames >= minimumOverlapFrames)
                fineCandidates.Add(candidate);
        }
        fineCandidates.Sort(delegate(MverPhaseCandidate left, MverPhaseCandidate right) {
            return right.FullMotionSimilarity.CompareTo(left.FullMotionSimilarity);
        });
        return fineCandidates.ToArray();
    }
}

File: src/tools/mver_phase_metrics.c
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <objbase.h>

#include "mver_phase_frames.h"

#include <locale.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>

enum { REGION_COUNT = 4 };

typedef struct Score {
    long long absolute[REGION_COUNT];
    long long samples[REGION_COUNT];
} Score;

typedef struct MotionScore {
    double dot[REGION_COUNT];
    double first_squared[REGION_COUNT];
    double second_squared[REGION_COUNT];
} MotionScore;

typedef struct Candidate {
    int offset;
    double seconds;
    int overlap;
    double similarity[REGION_COUNT];
    double motion[REGION_COUNT];
} Candidate;

static int in_region(int region, int x, int y, int width, int height) {
    double nx = x / (double)width, ny = y / (double)height;
    if (region == 0) return 1;
    if (region == 1) return nx >= .27 && nx <= .74 && ny >= .24 && ny <= .67;
    if (region == 2) return ny <= .36 || ((nx <= .30 || nx >= .71) && ny <= .70);
    return nx >= .16 && nx <= .58 && ny >= .58 && ny <= .92;
}

static void add_pair(Score *score, const Frame *first, const Frame *second,
    int spatial_stride) {
    int gx, gy, region;
    for (gy = 0; gy < first->grid_height; gy += spatial_stride)
    for (gx = 0; gx < first->grid_width; gx += spatial_stride) {
        const unsigned char *a = first->bgr +
            ((size_t)gy * first->grid_width + gx) * 3;
        const unsigned char *b = second->bgr +
            ((size_t)gy * second->grid_width + gx) * 3;
        int first_max = max(a[0], max(a[1], a[2]));
        int second_max = max(b[0], max(b[1], b[2]));
        int absolute, x, y;
        if (first_max <= 4 && second_max <= 4) continue;
        absolute = abs((int)a[0] - b[0]) + abs((int)a[1] - b[1]) +
            abs((int)a[2] - b[2]);
        x = gx * PIXEL_STEP; y = gy * PIXEL_STEP;
        for (region = 0; region < REGION_COUNT; ++region) if (in_region(region,
            x, y, first->width, first->height)) {
            score->absolute[region] += absolute;
            ++score->samples[region];
        }
    }
}

static void add_motion_pair(MotionScore *score, const Frame *first_previous,
    const Frame *first, const Frame *second_previous, const Frame *second,
    int spatial_stride) {
    int gx, gy, channel, region;
    for (gy = 0; gy < first->grid_height; gy += spatial_stride)
    for (gx = 0; gx < first->grid_width; gx += spatial_stride) {
        size_t offset = ((size_t)gy * first->grid_width + gx) * 3;
        int x = gx * PIXEL_STEP, y = gy * PIXEL_STEP;
        for (channel = 0; channel < 3; ++channel) {
            double first_delta = first->bgr[offset + channel] -
                first_previous->bgr[offset + channel];
            double second_delta = second->bgr[offset + channel] -
                second_previous->bgr[offset + channel];
            if (first_delta == 0.0 && second_delta == 0.0) continue;
            for (region = 0; region < REGION_COUNT; ++region) if (in_region(
                region, x, y, first->width, first->height)) {
                score->dot[region] += first_delta * second_delta;
                score->first_squared[region] += first_delta * first_delta;
                score->second_squared[region] += second_delta * second_delta;
            }
        }
    }
}

static double similarity(const Score *score, int region) {
    return score->samples[region] == 0 ? 0.0 : 100.0 * (1.0 -
        score->absolute[region] / (score->samples[region] * 3.0 * 255.0));
}

static double motion_similarity(const MotionScore *score, int region) {
    double denominator = sqrt(score->first_squared[region] *
        score->second_squared[region]);
    double correlation;
    if (denominator <= 0.0) return 0.0;
    correlation = score->dot[region] / denominator;
    correlation = max(-1.0, min(1.0, correlation));
    return 50.0 * (correlation + 1.0);
}

static Candidate measure(const Frame *mver, size_t mver_count,
    const Frame *native, size_t native_count, int offset,
    int interval_milliseconds, int spatial_stride) {
    int mver_start = max(0, -offset), native_start = max(0, offset);
    int overlap = min((int)mver_count - mver_start,
        (int)native_count - native_start);
    Score score = {{0}, {0}};
    MotionScore motion = {{0}, {0}, {0}};
    Candidate candidate;
    int frame, region;
    memset(&candidate, 0, sizeof(candidate));
    for (frame = 0; frame < overlap; frame += 2)
        add_pair(&score, &mver[mver_start + frame],
            &native[native_start + frame], spatial_stride);
    for (frame = 1; frame < overlap; ++frame)
        add_motion_pair(&motion, &mver[mver_start + frame - 1],
            &mver[mver_start + frame], &native[native_start + frame - 1],
            &native[native_start + frame], spatial_stride);
    candidate.offset = offset;
    candidate.seconds = offset * interval_milliseconds / 1000.0;
    candidate.overlap = overlap;
    for (region = 0; region < REGION_COUNT; ++region) {
        candidate.similarity[region] = similarity(&score, region);
        candidate.motion[region] = motion_similarity(&motion, region);
    }
    return candidate;
}

static int compare_candidates(const void *left, const void *right) {
    const Candidate *a = (const Candidate *)left, *b = (const Candidate *)right;
    if (a->motion[0] < b->motion[0]) return 1;
    if (a->motion[0] > b->motion[0]) return -1;
    return a->offset < b->offset ? -1 : a->offset > b->offset;
}

static void print_candidate(const Candidate *candidate, int comma) {
    printf("  {\"NativeOffsetFrames\":%d,\"NativeOffsetSeconds\":%.17g,"
        "\"OverlapFrames\":%d,\"FullSimilarity\":%.17g,"
        "\"FaceSimilarity\":%.17g,\"HairSimilarity\":%.17g,"
        "\"HandSimilarity\":%.17g,\"FullMotionSimilarity\":%.17g,"
        "\"FaceMotionSimilarity\":%.17g,\"HairMotionSimilarity\":%.17g,"
        "\"HandMotionSimilarity\":%.17g}%s\n", candidate->offset,
        candidate->seconds, candidate->overlap, candidate->similarity[0],
        candidate->similarity[1], candidate->similarity[2],
        candidate->similarity[3], candidate->motion[0], candidate->motion[1],
        candidate->motion[2], candidate->motion[3], comma ? "," : "");
}

int wmain(int argc, wchar_t **argv) {
    PathList mver_paths = {0}, native_paths = {0};
    Frame *mver = NULL, *native = NULL;
    Candidate *coarse = NULL, *fine = NULL;
    size_t coarse_count = 0, fine_count = 0, index;
    int max_lag, minimum_overlap, interval_ms, offset, result = 1;
    int selected_offsets[8], selected_count = 0;
    if (argc < 3 || argc > 6) {
        fwprintf(stderr, L"usage: mver-phase-metrics MVER_DIRECTORY "
            L"NATIVE_DIRECTORY [max-lag-frames] [minimum-overlap-frames] "
            L"[interval-ms]\n");
        return 2;
    }
    setlocale(LC_NUMERIC, "C");
    max_lag = argc > 3 ? _wtoi(argv[3]) : 30;
    minimum_overlap = argc > 4 ? _wtoi(argv[4]) : 2;
    interval_ms = argc > 5 ? _wtoi(argv[5]) : 17;
    if (max_lag < 0 || minimum_overlap < 2) return 2;
    if (FAILED(CoInitializeEx(NULL, COINIT_MULTITHREADED))) return 1;
    if (!phase_collect_paths(argv[1], &mver_paths) ||
        !phase_collect_paths(argv[2], &native_paths) ||
        mver_paths.count < (size_t)minimum_overlap ||
        native_paths.count < (size_t)minimum_overlap) goto done;
    mver = phase_load_frames(&mver_paths);
    native = phase_load_frames(&native_paths);
    if (!mver || !native) goto done;
    coarse = (Candidate *)malloc((size_t)(max_lag * 2 + 1) * sizeof(*coarse));
    fine = (Candidate *)malloc(8 * sizeof(*fine));
    if (!coarse || !fine) goto done;
    for (offset = -max_lag; offset <= max_lag; ++offset) {
        Candidate candidate = measure(mver, mver_paths.count, native,
            native_paths.count, offset, interval_ms, 3);
        if (candidate.overlap >= minimum_overlap) coarse[coarse_count++] = candidate;
    }
    if (!coarse_count) goto done;
    qsort(coarse, coarse_count, sizeof(*coarse), compare_candidates);
    for (index = 0; index < min((size_t)7, coarse_count); ++index)
        selected_offsets[selected_count++] = coarse[index].offset;
    {
        int found_zero = 0;
        for (index = 0; index < (size_t)selected_count; ++index)
            if (selected_offsets[index] == 0) found_zero = 1;
        if (!found_zero) selected_offsets[selected_count++] = 0;
    }
    for (index = 0; index < (size_t)selected_count; ++index) {
        Candidate candidate = measure(mver, mver_paths.count, native,
            native_paths.count, selected_offsets[index], interval_ms, 1);
        if (candidate.overlap >= minimum_overlap) fine[fine_count++] = candidate;
    }
    qsort(fine, fine_count, sizeof(*fine), compare_candidates);
    printf("[\n");
    for (index = 0; index < fine_count; ++index)
        print_candidate(&fine[index], index + 1 < fine_count);
    printf("]\n");
    result = 0;
done:
    free(coarse); free(fine);
    phase_free_frames(mver, mver_paths.count);
    phase_free_frames(native, native_paths.count);
    phase_free_paths(&mver_paths); phase_free_paths(&native_paths);
    CoUninitialize();
    if (result) fwprintf(stderr, L"phase analysis failed\n");
    return result;
}
